File: adventOfCode/05/solution.js
var fs = require('fs');

const polymer = fs.readFileSync('input.txt', 'ascii').replace('\n', '');

function collapse(polymer, drop) {
  let final_polymer = [];
  polymer.replace(new RegExp(drop, 'gi'), '').split('').forEach((a) => {
    final_polymer.push(a);
    while (reacts(...final_polymer.slice(a.length - 3))) {
      final_polymer.pop();
      final_polymer.pop();
    }
  });
  return final_polymer.join('');
}

function reacts(a, b) {
  if (!a || !b) {return false;}
  return (a != b) && (a.toLowerCase() === b.toLowerCase());
}


// Part One


const collapsed = collapse(polymer);
console.log(collapsed.length);


// Part Two


const alphabet = [...Array(26).keys()]
  .map((x) => String.fromCharCode(x + 97));
console.log(
  Math.min(...alphabet.map(
    (a) => collapse(collapsed, a).length
  ))
);

File: adventOfCode/05/solution.py
with open('input.txt', 'r') as inputfile:
    polymer = inputfile.read().replace('\n', '')


def collapse(polymer, drop=''):
    final_polymer = []
    for unit in polymer.replace(drop.lower(), '').replace(drop.upper(), ''):
        final_polymer.append(unit)
        last_two = final_polymer[-2:]
        while len(final_polymer) >= 2 and react(*last_two):
            final_polymer[-2:] = []
            last_two = final_polymer[-2:]
    return ''.join(final_polymer)


def react(a, b):
    return a != b and a.lower() == b.lower()


# Part One


collapsed = collapse(polymer)
print(len(collapsed))


# Part Two

alphabet = [chr(i) for i in range(65, 65+26)]
print(min(len(collapse(collapsed, alpha)) for alpha in alphabet))
